### PR TITLE
Add curves CSV loader

### PR DESCRIPTION
**Pull request by jodastephen**
Move from strata-examples.
Fix and enhance curve calibration example to use quotes CSV loader,

### CODE REVIEW

File: examples/src/main/java/com/opengamma/strata/examples/finance/CalibrationCheckExample.java [hunk 1]
```diff
@@ -89,15 +93,20 @@
...
+  private static final String CURVE_GROUP_CONFIG_FILE_NAME = PATH_CONFIG + "curves/USD-DSCON-LIBOR3M.xml";
   /**
    * The location of the market quotes file.
    */
-  private static final String MARKET_QUOTES_FILE_NAME = PATH_CONFIG + "market_quotes_usd.xml";
+  private static final ResourceLocator QUOTES_RESOURCE =
+      ResourceLocator.of(ResourceLocator.FILE_URL_PREFIX + PATH_CONFIG + "quotes/quotes.csv");
+
+  static {
+    Reflection.initialize(LoaderUtils.class);
```
Comment (cjkent) [inline examples/src/main/java/com/opengamma/strata/examples/finance/CalibrationCheckExample.java]: Why is this necessary? It doesn't smell good.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/LoadedCurveSettings.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  /**
+   * Obtains an instance.
+   * 
+   * @param curveName  the curve name
+   * @param yValueType  the value type
+   * @param dayCount  the day count
+   * @param interpolator  the interpolator
+   * @param extrapolatorLeft  the left extrapolator
+   * @param extrapolatorRight  the right extrapolator
+   * @return the curve node
```
Comment (cjkent) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/LoadedCurveSettings.java]: Copy / paste error

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoader.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+   * Loads one or more CSV format curve files for a specific date.
+   * <p>
+   * Only those quotes that match the specified date will be loaded.
+   * <p>
+   * If the files contain a duplicate entry an exception will be thrown.
+   * 
+   * @param groupsResource  the curve groups CSV resource
+   * @param settingsResource  the curve settings CSV resource
+   * @param curvesResources  the CSV resources for curves
+   * @param marketDataDate  the curve date to load
```
Comment (cjkent) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoader.java]: The parameter Javadoc aren't in the same order as the parameters

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoader.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+      Set<RateCurveId> curveUses = curveGroups.computeIfAbsent(curveName, k -> new HashSet<RateCurveId>());
+      curveUses.add(curveId);
+    }
+    return curveGroups;
+  }
+
+  // parses the identifier
+  private static RateCurveId createRateCurveId(CurveGroupName curveGroup, String curveTypeStr, String referenceStr) {
+    // discount and forward curves are supported
+    if ("forward".equals(curveTypeStr.toLowerCase())) {
```
Comment (cjkent) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoader.java]: `forward` and `discount` should probably be constants.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoader.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+      return DiscountCurveId.of(ccy, curveGroup);
+
+    } else {
+      throw new IllegalArgumentException(Messages.format("Unsupported curve type: {}", curveTypeStr));
+    }
+  }
+
+  //-------------------------------------------------------------------------
+  // loads a single curves CSV file
+  private static Map<LoadedCurveKey, Curve> loadSingle(
```
Comment (cjkent) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoader.java]: Comment to say that `requestedDate` can be null and what that implies

File: modules/loader/src/test/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoaderTest.java [hunk 6]
```diff
@@ -64,6 +73,11 @@
   // tolerance
   private static final double TOLERANCE = 1.0E-4;
 
+  static {
+    LoaderUtils.findIndex("USD-LIBOR-3M");
+//    Reflection.initialize(LoaderUtils.class);
```
Comment (cjkent) [inline modules/loader/src/test/java/com/opengamma/strata/loader/csv/RatesCurvesCsvLoaderTest.java]: Delete the commented-out line

**Comment by cjkent** [comment]
That's all from me

File: examples/src/main/java/com/opengamma/strata/examples/finance/CalibrationCheckExample.java (hunk 1 above)
Comment (jodastephen) [inline examples/src/main/java/com/opengamma/strata/examples/finance/CalibrationCheckExample.java]: This is a workaround until a Joda-Beans fix happens. I'll comment
